govuk-one-login/ipv-cri-kbv-hmrc-api
# LIME-1197 Add Error handling to unhappy paths.

## Proposed changes

### What changed

	- Template
	- Remove UserAgent, QuestionsUrl and AnswersUrl
	- Add reading the above from an optional parameter prefix
	- Remove BearerTokenName from Template (replaced by otg-token lambda)
	- Deleted and removed execute-state-machine lambda from Template
	- Deleted and removed CheckSessionStateMachine from Template
	- Deleted and removed CurrentTimeFunction from Template
	- Deleted and removed TimeFunction from Template

	- Types
	- add SessionItem as an interface (common-type.js)
	- add PersonIdentityItem as an interface (common-type.js)

	- Add dynamo-unmarshall lambda function
	- - To achieve the capablity of the high level dynamo client.

	- Fetch Questions State Machine
	- Unmarshall SessionItem from the DynamoDB response
	- Validate the SessionItem has the expected fields.
	- Replace use of raw dynamo response with SessionItem
	- Unmarshall PersonIdentityItem from the DynamoDB response
	- Handle errors in the OTG and Fetch Questions lambda seperately
	- FetchQuestionInputs to expect PersonIdentityItem
	- Update Fetch Questions VTL
	- Correct test to expect a 401 from an invalid sessionId or missing
	- Change default response status code to 500

	- Get Question State Machine
	- Add getQuestionState to final output to determine status code
	- getQuestionState = "Question", 200 + question key/ info returned
	- getQuestionState = "NoMoreQuestions", 204
	- Session is checked to have an associated sesionItem and personIdenity with Nino
	- Uses Raw Dynamo Paths as unmarshalling is not required (no lambda)
	- Update Get Question VTL
	- VTL Conditionally output info, current, previous tax years.
	- Add Fail States
	- - SessionId missing or not found
	- - SessionItem not valid
	- - PersonIdentityItem not found
	- - NINO was not present in shared claims

	- Post Answer State Machine
	- Adjust States to remove overalapping flow in sucess paths (Has the user already answered all questions)
	- Check the incomming sessionId is assoicated with a sessionItem
	- Check return of answer-validation lambda
	- Renamed "Parse input" to "Remap Input Data (answeredQuestionKey and inputAnswer)"
	- Check return of otgtoken lambda
	- Check return of submit-answer lambda
	- Check return of CreateAuthCode lambda
	- Remove Create Parameters State Location and write parameter values into a known location for each parameter (.parameters.parametername.value) as with the other state machines
	- Update Post Answer VTL
	- Add Fail States
	- - SessionId missing or not found
	- - SessionItem not valid
	- - Answers failed validation
	- - OTG Token Lambda responded unexpectedly
	- - Submit Answers Lambda responded unexpectedly
	- - CreateAuthCodeFunction responded unexpectedly

	- Issue Credential State Machine
	- Default to 500 Error
	- Replace use of raw dynamo response with SessionItem in IssueCredentialHandler
	- Move parameters to .parameters
	- Assert VC is present (.jwt) as final check
	- Add success state
	- Add Fail states -
	- - Bearer Token missing
	- - Invalid Bearer Token
	- - SessionItem not valid
	- - PersonIdentityItem not found
	- - Issuecredential Lambda responded unexpectedly
	- - JwtSignerFunction Lambda responded unexpectedly

	- Issue Credential Lambda
	- Add IssueCredentialInputs to control input parsing
	- Use parameters, sessionItem and personIdentityItem in IssueCredentialInputs.
	- CredentialSubjectBuilder switch to common-types
	- Return VC Body in { vcBody:value} to enable asserting correct return in state machine
	- Remove StepFunctionsExecutionPolicy

	- JwtSignerHandler
	- Add top level catch
	- Add completion metric
	- Align tests to catch errors as lambda output
	- return VC in { jwt:value} to enable asserting correct return in state machine

	- Audit Service
	- Use SessionItem instead of raw dynamo response
	- Use PersonIdentityItem instead of custom types

	- Integration test
	- update fetch-questions tests to account for changes

	- Feature tests
	- Aligned with status code changes
	- Update feature-test to correct/add logging to post answers

	- Readme
	- Add command to run the pre-merge tests
	- Update command to run the integration tests to include the parameter-prefix

### Why did it change

To improve error handling thought the backend API

### Issue tracking



- [LIME-1197](https://govukverify.atlassian.net/browse/LIME-1197)

### Other considerations
